overwatch/src/services/runner.rs

```diff
@@ -0,0 +1,364 @@
+            status_handle.clone(),
+            overwatch_handle.clone(),
+            settings_updater.clone(),
+            state_updater,
+            lifecycle_handle.clone(),
+        );
+
+        Self {
+            service_resources,
+            state_handle,
+            lifecycle_handle,
+            overwatch_handle,
+            settings_updater,
+            status_handle,
+            relay,
+            relay_buffer_size,
+        }
+    }
+}
+
+impl<Message, Settings, State, StateOp, RuntimeServiceId>
+    ServiceRunner<Message, Settings, State, StateOp, RuntimeServiceId>
+where
+    Message: 'static + Send,
+    Settings: Clone + 'static + Sync + Send,
+    State: ServiceState<Settings = Settings> + Clone + Send + Sync + 'static,
+    <State as ServiceState>::Error: Display,
+    StateOp: StateOperator<State = State> + Send + 'static,
+    RuntimeServiceId: 'static + Clone + Send,
+{
+    /// Spawn the service main loop and handle its lifecycle.
+    ///
+    /// # Returns
+    ///
+    /// A tuple containing the service id and the lifecycle handle, which allows
+    /// to manually abort the execution.
+    ///
+    /// # Errors
+    ///
+    /// If the service cannot be initialized properly with the retrieved state.
+    pub fn run<Service>(
+        self,
+    ) -> ServiceRunnerHandle<Message, Settings, State, StateOp, RuntimeServiceId>
+    where
+        Service: ServiceCore<RuntimeServiceId, Settings = Settings, State = State, Message = Message>
+            + 'static,
+        StateOp: Clone,
+    {
+        let service_handle = ServiceHandle::from(&self);
+        let runtime = self.service_resources.overwatch_handle.runtime().clone();
+        let runner_join_handle = runtime.spawn(self.run_::<Service>());
+
+        ServiceRunnerHandle {
+            service_handle,
+            runner_join_handle,
+        }
+    }
+
+    async fn run_<Service>(self)
+    where
+        Service: ServiceCore<RuntimeServiceId, Settings = Settings, State = State, Message = Message>
+            + 'static,
+        StateOp: Clone,
+    {
+        let Self {
+            service_resources,
+            state_handle,
+            mut lifecycle_handle,
+            relay,
+            relay_buffer_size,
+            status_handle,
+            ..
+        } = self;
+
+        let Relay {
+            inbound,
+            outbound: _,
+            consumer_sender,
+            consumer_receiver,
+        } = relay;
+
+        let runtime = service_resources.overwatch_handle.runtime().clone();
+        let mut service_task_handle: Option<_> = None;
+        let mut state_handle_task_handle: Option<_> = None;
+
+        let mut inbound_relay = Some(inbound);
+
+        while let Some(lifecycle_message) = lifecycle_handle.next().await {
+            match lifecycle_message {
+                LifecycleMessage::Start(sender) => {
+                    if !status_handle.borrow().is_startable() {
```

Comment:
Do you think we could encode the fields such as channels and related that are only present when the service is running as an enum? So we are forced by the compiler to check what state we are in to access those resources? I feel it would be better than having a boolean tell us what to do, which we could also "miss" in a future refactoring.

Reply:
Yes, a state pattern would fit well. Nice catch @ntn-x2

Reply:
I understand what you mean but I'm failing to completely grasp it. Is it close to what I suggested [here](https://github.com/logos-co/Overwatch/pull/79/files#r2091067925) as the second alternative?

Reply:
> What you proposed is probably part of something larger in scope that I am suggesting. What I am suggesting is that stuff that is part of a service only when the service is running, might (I am saying might because I am not sure it's possible) be part of an enum, with one variant for when the service is running and that contains all the stuff that service has access to.

Better to have a struct that morphs its type state (do you remember the interview @ntn-x2 ? 😉 ). Not sure if its possible here though. But would be better.

```rust
struct Running;
struct Stopped;

struct StatusHandle<State> {}
impl StatusHandle<Running> {
    pub fn stop(self) -> Self<Stop>;
}

impl StatusHandle<Stopped> {
    pub fn run(self) -> Self<Running>;
}

```